tmt/steps/context/pidfile.py

```diff
@@ -0,0 +1,296 @@
+"""
+Pidfile handling.
+
+tmt must make sure running a script must allow for multiple external
+factors: the test timeout, interactivity, reboots and `tmt-reboot`
+invocations. tmt must present consistent info on what is the PID to
+kill from `tmt-reboot`, and where to save additional reboot info.
+
+.. note::
+
+    Historically, this is mostly visible with tests, but the concept
+    applies to all restartable actions on the guest: ``prepare/shell``
+    scripts, ``finish/ansible`` playbooks, and so on.
+
+To achieve these goals, tmt uses two wrappers, the inner and the outer
+one. The inner one wraps the actual action, e.g. a test script as defined
+in test metadata, the outer one then runs the inner wrapper while
+performing the necessary setup and accounting. tmt invokes the outer
+wrapper which then invokes the inner wrapper which then invokes the action.
```

Comment:
Moved from https://github.com/teemtee/tmt/pull/4282/files#r2510552720

---

Could use a bit more clarity in that the inner wrapper is the test field and the outer contains the setup defined by tmt.